virtual_ecosystem/core/configuration.py

```diff
@@ -0,0 +1,151 @@
+
+
+class ModelConfigRoot(ConfigRoot):
+    """Root configuration class for models.
+
+    This model provides a common Pydantic base class that must be used to define
+    the root configuration class of a Virtual Ecosystem model. Each model must define a
+    single class inheriting from :class:`ModelConfigRoot` in a ``model_config.py``
+    submodule. The file can then include other :class:`ModelConfigSection` classes that
+    are used as nested fields within the root configuration but can be only one
+    :class:`ModelConfigRoot` class per model. This base model sets common shared
+    attributes across models: currently just the shared ``static`` option.
+    """
+
+    static: bool = False
+    """The model static mode setting."""
+
+
+class ModelConfigSection(ConfigRoot):
+    """Section configuration class for models.
+
+    This model provides a common base class for nested subsections within model
+    configurations. The base model inherits shared configuration settings from the
+    :class:`ConfigRoot` class.
+
+    This is functionally an alias of :class:`ConfigRoot` and is used purely to
+    differentiate nested sections from the core config root.
+    """
+
+
+def placeholder_validator(path: str) -> str:
+    """A custom validator to reject "<PLACEHOLDER>" when loading file paths."""
+    if path == "<PLACEHOLDER>":
+        raise ValueError("Path placeholder value in configuration.")
+
+    return path
+
+
+FILEPATH_PLACEHOLDER: TypeAlias = Annotated[
+    FilePath,
+    Field(default=Path("<PLACEHOLDER>")),
+    BeforeValidator(placeholder_validator),
+]
+"""Pydantic type that provides a default '<PLACEHOLDER>' text for writing configuration
+templates, but screens input before the standard validation to refuse unreplaced
+placeholder values. The type then uses FilePath, which validates that a path actually
+exists on the file system. The field does not set ``validate_defaults`` so the
+placeholder value can be written despite not being an existing file.
+
+.. TODO: Fix autodoc
+    This generates a bizarre set of autodoc link failures that try and build links from
+    the text elements from the Annotated pattern. Currently tackled using nitpick 
+    ignore.
+"""
+
+
+class ModelConfigHTMLTable:
+    """Class to render the fields in a ModelConfig class as an HTML Table.
+
+    This class is a helper function for use in documenting model configurations. It
+    takes a model configuration class and then iterates over model fields, recursing
+    into sub-models within the fields, to generate a simple HTML table showing the
+    config sections and then the description and defaults of each setting.
+
+    Args:
+        model_name: The name of the model as it would appear in a configuration file.
+        config_object: A ModelConfig instance
+    """
+
+    def __init__(
+        self,
+        model_name: str,
+        config_object: type[ModelConfigRoot] | type[ModelConfigSection],
```

Comment:
If it can be either one of these... then it must be a `ConfigRoot`, right?

```suggestion
        config_object: type[ConfigRoot],
```